test.py

```diff
@@ -842,33 +842,36 @@ async def run(self, options: argparse.Namespace) -> Test:
-                self.args.insert(0, "--host={}".format(cluster.endpoint()))
-                self.is_before_test_ok = True
-                cluster.take_log_savepoint()
-                status = await run_test(self, options)
-                cluster.after_test(self.uname)
-                self.is_after_test_ok = True
-                self.success = status
-            except Exception as e:
-                self.server_log = cluster.read_server_log()
-                self.server_log_filename = cluster.server_log_filename()
-                if self.is_before_test_ok is False:
-                    print("Test {} pre-check failed: {}".format(self.name, str(e)))
-                    print("Server log of the first server:\n{}".format(self.server_log))
-                    # Don't try to continue if the cluster is broken
-                    cm.discard()
-                if self.is_after_test_ok is False:
-                    print("Test {} post-check failed: {}".format(self.name, str(e)))
-                    print("Server log of the first server:\n{}".format(self.server_log))
-                    logger.info(f"Discarding cluster after failed test %s...", self.name)
-                    await self.suite.clusters.steal()
-                    await cluster.stop()
-                    await cluster.release_ips()
-                    cm.discard()
+        cluster = await self.suite.clusters.get()
+        try:
+            cluster.before_test(self.uname)
+            logger.info("Leasing Scylla cluster %s for test %s", cluster, self.uname)
+            self.args.insert(0, "--host={}".format(cluster.endpoint()))
+            self.is_before_test_ok = True
+            cluster.take_log_savepoint()
+            status = await run_test(self, options)
+            cluster.after_test(self.uname)
+            self.is_after_test_ok = True
+            self.success = status
+        except Exception as e:
+            self.server_log = cluster.read_server_log()
+            self.server_log_filename = cluster.server_log_filename()
+            if self.is_before_test_ok is False:
+                print("Test {} pre-check failed: {}".format(self.name, str(e)))
+                print("Server log of the first server:\n{}".format(self.server_log))
+                logger.info(f"Discarding cluster after failed start for test %s...", self.name)
+                await self.suite.clusters.steal()
+                await cluster.stop()
+                await cluster.release_ips()
+            elif self.is_after_test_ok is False:
+                print("Test {} post-check failed: {}".format(self.name, str(e)))
+                print("Server log of the first server:\n{}".format(self.server_log))
+                logger.info(f"Discarding cluster after failed test %s...", self.name)
+                await self.suite.clusters.steal()
+                await cluster.stop()
+                await cluster.release_ips()
+        else:
+            await self.suite.clusters.put(cluster)
```

Comment:
By changing the indentation you moved the leak from one path to another. Now the happy path isn't leaking, but the path you just removed the `put` from (exception raised, but cluster checks OK) is leaking.

Reply:
In theory either of the conditions should've failed as they are both set to `False` before a test starts. But yes, it makes sense to just stop the cluster for all exceptions. And code is more compact.

Reply:
Oh, okay. I thought that it's possible for a test to fail without messing up the cluster, and then the cluster can be reused even after a failure. That's how the code before the patch behaves.